text/0000-no-entry.md

```diff
@@ -0,0 +1,149 @@
+symbol may represent initial boot code, applications running under an operating
+system but with unusual startup requirements, or programs creating or running
+in sandbox environments.
+
+The target toolchain will commonly supply an entry-point symbol (e.g.
+`_start`) by default. As a result, attempting to define an entry-point symbol
+typically results in a link error. Addressing this currently requires linking
+manually with an toolchain-specific argument like `-nostartfiles`.
+
+This proposal introduces a top-level attribute `no_entry`, which will cause the
+Rust toolchain to omit the entry-point symbol. Specifying `#![no_entry]` allows
+(and requires) the user to supply such a symbol.
+
+(Please note that this is entirely distinct from the Rust-supplied startup hook
+`#[start]`.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Specifying the attribute `#![no_entry]` at the top level of a Rust binary crate
+will cause Rust to omit the entry-point symbol when compiling and linking the
+binary. When linking, Rust will pass any necessary target-specific arguments to
+the toolchain to omit the target's entry-point code, such as `-nostartfiles`.
+
+Binaries built with this attribute set will typically not link unless the user
+supplies their own entry-point symbol.
+
+Libraries cannot use this attribute; attempting to do so will produce a
+compile-time error stating that `no_entry` only applies to binaries.
+
+A binary built with this attribute set could supply the entry-point symbol
+itself, or have the entry-point symbol supplied by a library crate or by a
+non-Rust library.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Until stabilized, the `no_entry` attribute will require a feature-gate named
+`no_entry`.
+
+The `no_entry` attribute may be specified as part of a `cfg_attr` declaration,
+to omit the entry point in some configurations but not others.
+
+Specifying the `no_entry` attribute on anything other than the top-level module
+of a binary crate will produce an error (`warning: crate-level attribute should
+be in the root module`).
+
+Declaring an entry-point symbol will typically require a declaration with
+`#[no_mangle] pub extern "C"`.
+
+To implement `no_entry`, the compiler can have a default linker argument based
+on the type of linker in use, such as `-nostartfiles` for GCC or clang/LLD.
+Specific targets which require more specific options can provide those options
+via the target configuration structures.
+
+Some targets may be able to support invoking the C library without running the
+startup code provided by the C library; other targets may not support this, or
+have limitations on such support. Targets that do not support this should
+document the target-specific limitations of `no_entry` for their target, may
+require `no_std` in addition to `no_entry`, or may not support `no_entry` at
```

Comment:
This also entails that supporting `no_entry` is optional for a Rust compiler in terms of static semantics and that the attribute has primarily implementation defined semantics, right? Would be good to spell that out if so.

Reply:
> This also entails that supporting no_entry is optional for a Rust compiler in terms of static semantics and that the attribute has primarily implementation defined semantics, right?

More than "optional", I understand this document to specify the behavior of `#![no_entry]` as implementation-defined, meaning that the implementation _must_ document what the behavior is.

Often, the language spec constraints the sets of behaviors that can be implemented for implementation-defined behavior, and I believe that it is in the "spirit" of this document to constraint the behaviors that a Rust implementation can provide to:

* compilation error if the implementation does not support `#![no_entry]` for a target 
* compilation error if the implementation does support `#![no_entry]` for a target but using it requires `#![no_std]`
* #![no_entry] does not instantaneously trigger a compilation error, but is handled according to the semantics of the target, which must be documented by the implementation for the particular target according to the text above. This might mean that compilation might fail later, e.g., if some symbol is missing.

That is, a Rust implementation cannot just choose to completely ignore `#![no_entry]` and, e.g., not being even able to parse it. It must explicitly support it, however, what `#![no_entry]` actually then does will depend on the implementation and the target.